File: dp/subsequence_subarray/maximum_subarray/max_subarray.go
package maximumsubarray

func maxSubArray(nums []int) int {
	n := len(nums)
	// dp[i]表示0-i,以i结尾的最大子数组和，子数组与子序列不同的是，子数组必须是连续的
	dp := make([]int, n)
	dp[0] = nums[0]
	ans := nums[0]
	for i := 1; i < n; i++ {
		dp[i] = max(dp[i-1]+nums[i], nums[i])
		ans = max(ans, dp[i])
	}
	return ans
}

func max(a, b int) int {
	if a < b {
		return b
	}
	return a
}

File: greedy/max_subarray_sum/MaxSubarray.java
package greedy.max_subarray_sum;

/**
 * leetcode 53 middle 最大子数组和
 * 
 * 给你一个整数数组 nums，请你找出一个具有最大和的连续子数组（子数组最少包含一个元素），返回其最大和。
 * 子数组 是数组中的一个连续部分。
 */
public class MaxSubarray {
    public int maxSubArray(int[] nums) {
        int n = nums.length;
        int[] dp = new int[n];
        dp[0] = nums[0];

        int ans = Integer.MIN_VALUE;

        for (int i = 1; i < n; i++) {
            dp[i] = Math.max(dp[i - 1] + nums[i], nums[i]);
            ans = Math.max(ans, dp[i]);
        }   

        return ans;
    }
}
